### PR TITLE
Retrieving whole line basing on offset

### PR DESCRIPTION
**Pull request by konefalg**

### CODE REVIEW

File: whylog/log_reader/read_utils.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        after = cls._expand_left(fd, offset, buf_size)
+        before = cls._expand_right(fd, offset, buf_size)
+        if not before:
+            raise Exception("Empty file!")
+        if not after:
+            raise Exception("Offset bigger than file size!")
+        return before[-1] + after[0]
+
+    @classmethod
+    def get_line_containing_offset(cls, fd, offset, buf_size):
```
Comment (ppolewicz) [inline whylog/log_reader/read_utils.py:55]: the caller probably needs to know the offsets of line start and line end to bisect properly

File: whylog/log_reader/read_utils.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    @classmethod
+    def _join_results(cls, first_part, second_part):
+        if not first_part:
+            if not second_part:
+                return []
+            return second_part
+        if not second_part:
+            return first_part
+        return first_part[:-1] + ["".join([first_part[-1], second_part[0]])] + second_part[1:]
```
Comment (ppolewicz) [inline whylog/log_reader/read_utils.py]: In the join use a tuple instead of list

File: whylog/tests/tests_log_reader/test_read.py [hunk 3]
```diff
@@ -29,6 +32,19 @@ def setUpClass(cls):
...
+        assert all(
+            [
+                [
+                    (('aaa-%d-bbb' % (i // 10)), (i // 10) * 10, (i // 10) * 10 + 9
+                    ) for i in six.moves.range(100)
+                ] == [
+                    ReadUtils.get_line_containing_offset(fh, i, j) for i in six.moves.range(100)
+                ] for j in six.moves.range(1, 120, 7)
+            ]
+        )
```
Comment (ppolewicz) [inline whylog/tests/tests_log_reader/test_read.py]: `# yapf: disable`

File: whylog/tests/tests_log_reader/test_read.py [hunk 4]
```diff
@@ -29,6 +32,19 @@ def setUpClass(cls):
             )
         )  # yapf: disable
 
+    def test_getting_line_by_offset(self):
+        fh = open(TestPaths.get_file_path(AFewLinesLogParams.FILE_NAME))
+        assert all(
```
Comment (ppolewicz) [inline whylog/tests/tests_log_reader/test_read.py]: please rewrite this from list comprehension into standard `for` loops

File: whylog/tests/tests_log_reader/test_read.py [hunk 5]
```diff
@@ -29,6 +32,19 @@ def setUpClass(cls):
             )
         )  # yapf: disable
 
+    def test_getting_line_by_offset(self):
+        fh = open(TestPaths.get_file_path(AFewLinesLogParams.FILE_NAME))
```
Comment (ppolewicz) [inline whylog/tests/tests_log_reader/test_read.py]: `with open(...) as fh:`

File: whylog/log_reader/read_utils.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        while len(after) < 2:
+            lines = cls._read_split_lines(fd, position, buf_size)
+            if lines == ['']:
+                break
+            after = cls._join_results(after, lines)
+            position += buf_size
+        return after
+
+    @classmethod
+    def _expand_right(cls, fd, position, buf_size):
```
Comment (ppolewicz) [inline whylog/log_reader/read_utils.py]: please make a simple test which shows how this works

File: whylog/log_reader/read_utils.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            if position <= 0:
+                lines = cls._read_split_lines(fd, 0, position + buf_size)
+                before = cls._join_results(lines, before)
+                break
+            lines = cls._read_split_lines(fd, position, buf_size)
+            before = cls._join_results(lines, before)
+        return before
+
+    @classmethod
+    def _read_entire_line(cls, fd, offset, buf_size):
```
Comment (ppolewicz) [inline whylog/log_reader/read_utils.py:49]: please make a simple test which shows how this works

File: whylog/tests/tests_log_reader/test_read.py [hunk 8]
```diff
@@ -29,6 +32,19 @@ def setUpClass(cls):
             )
         )  # yapf: disable
 
+    def test_getting_line_by_offset(self):
```
Comment (ppolewicz) [inline whylog/tests/tests_log_reader/test_read.py]: please add a test which is less clever, but very easy to read

File: whylog/log_reader/exceptions.py [hunk 9]
```diff
@@ -13,3 +13,11 @@ def __str__(self):
         return 'Config did not match any log_type to the specified front_input: %s; cannot investigate' % (
             self._front_input
         )
+
+
+class ReadingError(LogReaderError):
```
Comment (ppolewicz) [inline whylog/log_reader/exceptions.py]: this should be two classes, one for empty file and one for offset bigger than file size

File: whylog/tests/tests_log_reader/test_read.py (hunk 8 above)
Comment (konefalg) [inline whylog/tests/tests_log_reader/test_read.py]: I've improved function readability. Please check it again and tell if another test is really necessary.

File: whylog/log_reader/read_utils.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            return first_part
+        return first_part[:-1] + ["".join((first_part[-1], second_part[0]))] + second_part[1:]
+
+    @classmethod
+    def _expand_after(cls, fd, position):
+        fd.seek(position)
+        line = fd.readline()
+        if not line:
+            raise OffsetBiggerThanFileSize(position)
+        return line.strip()
```
Comment (ppolewicz) [inline whylog/log_reader/read_utils.py]: whitespace can be meaningful for someone, you can't just drop part of the data like that. If there is a trailing newline or something, use offsets to deal with it

File: whylog/tests/tests_log_reader/test_read.py (hunk 8 above)
Comment (ppolewicz) [inline whylog/tests/tests_log_reader/test_read.py]: a file looking like this:

```
12
45
78
```

can be used to demonstrate methods from `ReadUtils` class. Either show how they work in tests, or document them - this is not a trivial code and it should be documented somehow. I'd prefer tests over a docstring explanation.

**Pull request merged**